Merge the jdk-21.0.9+3-mmtk branch into master

wks opened the pull request:
We merge the branch for OpenJDK 21 into `master`.  From now on, the `master` branch of the `mmtk-openjdk` repository will directly target OpenJDK 21.  The `jdk-21.0.9+3-mmtk` branch will be decommissioned, and will remain its current state (i.e. the state just before merging into `master`), and will only be looked up when any developers are interested in the history of the development of the OpenJDK binding.

The following things work:

-   We can run OpenJDK 21 with MMTk on x86_64, and it supports compressed oops.
-   CI tests for pull requests of the `mmtk-openjdk` binding work.
-   CI binding tests (including both minimal tests and extended tests) for pull requests of the `mmtk-core` repo work.

One important thing that doesn't work immediately is the **performance regression CI**.  It requires a change in the `ci-perf-kit` repo and also needs related changes in the CI scripts in `mmtk-core`.  This PR is for OpenJDK, and will not trigger the performance regression CI.  This leaves us a small window to update `mmtk-core` and `ci-perf-kit` after this PR is merged.

**This PR should not be merged with "squash merge"**.  Instead we should make a proper merge commit so that the `jdk-21.0.9+3-mmtk` branch will be considered as an ancestor of the `master` branch from Git's point of view.  GitHub will also consider the branch `jdk-21.0.9+3-mmtk` as "merged".

# Major changes compared to the OpenJDK 11 binding

We made the following changes in the mmtk-openjdk repo:

-   We made changes to reflect upstream changes in OpenJDK, including initialization, mutex API, roots, JIT compilers, etc.
-   We now support multiple architectures, including x86_64, ARM64 and RISC-V 64.  We changed the directory structure so that architecture-specific parts are held in separate files.
-   We updated the CI scripts to reflect related changes, such as removing the `-normal` part from the OpenJDK CONF string.

xinyu-zheng committed:
Make NoGC work

xinyu-zheng committed:
Fix initializer order

xinyu-zheng committed:
Match Klasses layout

xinyu-zheng committed:
Fix savengable seg fault

xinyu-zheng committed:
Make scavengable return true

xinyu-zheng committed:
Apply root scanning optimizations

xinyu-zheng committed:
Add missing roots

xinyu-zheng committed:
Fix undefined reference

xinyu-zheng committed:
Fix hidden virtual functions

xinyu-zheng committed:
Remove unnecessary set_native_thread_name

xinyu-zheng committed:
Fix parse_access cast

xinyu-zheng committed:
Fix making JVM_LIB_OUTPUTDIR

xinyu-zheng committed:
Remove explicit JNI roots scanning

xinyu-zheng committed:
Fix code cache roots scanning

xinyu-zheng committed:
Merge v0.12 into 17-0.12

xinyu-zheng committed:
Delete scratch directory

xinyu-zheng committed:
Fix fastdebug

xinyu-zheng committed:
Merge branch '17-0.12' of github.com:xinyu-zheng/mmtk-openjdk into 17-0.12

caizixian committed:
Fix store_at

caizixian committed:
Fix set_threadObj

caizixian committed:
Update jdk commit hash

caizixian committed:
Reflect changes in oops

caizixian committed:
Reflect changes in Mutex and Monitor

caizixian committed:
Fix includes

caizixian committed:
Fix c2 alloc sequence due to changes in allocation dprobe

caizixian committed:
Reflect changes in thread type

caizixian committed:
Fix warning

caizixian committed:
Add assertion when using lightweight threads with MMTk

caizixian committed:
Fix code marking

Wenyu says we don't do class unloading. This is fine.

caizixian committed:
Merge remote-tracking branch 'upstream/master' into jdk-20+15-mmtk-merge

caizixian committed:
Fix compiler errors

caizixian committed:
Have zero working with no barrier

caizixian committed:
Restore object barrier

caizixian committed:
Merge remote-tracking branch 'upstream/master' into jdk-20+15-mmtk-zero

caizixian committed:
Conditional compile CPU specific code

caizixian committed:
Merge pull request #1 from caizixian/jdk-20+15-mmtk-zero

Zero interpreter support + merge mmtk-openjdk 0.15

caizixian committed:
Update OpenJDK commit hash

caizixian committed:
Basic RISC-V fast path structure

caizixian committed:
Fix code cache roots relocation

caizixian committed:
RISC-V cross compilation

caizixian committed:
Implement eden_allocate

caizixian committed:
Merge pull request #2 from caizixian/jdk-20+16-mmtk-fastpath

Basic non-zero build support on riscv

caizixian committed:
Fix include

caizixian committed:
Always use the installed stable toolchain

caizixian committed:
Implement C1 write barrier stub

caizixian committed:
Refactor mmtkNoBarrier

caizixian committed:
Refactor mmtkObjectBarrier

caizixian committed:
Implement arraycopy_epilogue

caizixian committed:
Finish porting the object barrier for riscv

caizixian committed:
More assertions

caizixian committed:
fop interpreter working with no barrier fast path and basically nop slow path

Set mmtk-core FULL_NURSERY_GC to true as well

java -XX:+UseThirdPartyHeap -XX:MetaspaceSize=1G -XX:+UnlockDiagnosticVMOptions  -Xint -Xmx60M -Xms60M -jar /mnt
/dacapo/dacapo-evaluation-git-f480064.jar fop

caizixian committed:
fop interpreter with no barrier fast path but with slow path call (only pass the first argument)

Set mmtk-core FULL_NURSERY_GC to true, working, but false, not working

java -XX:+UseThirdPartyHeap -XX:MetaspaceSize=1G -XX:+UnlockDiagnosticVMOptions  -Xint -Xmx60M -Xms60M -jar /mnt/dacapo/dacapo-evaluation-git-f480064.jar fop

caizixian committed:
No barrier fast path but with slow path call (only pass the first argument)

fop working with interpreter with no mmtk-core change

MMTK_PLAN=GenCopy /mnt/jdk/build/linux-riscv64-server-slowdebug/jdk/bin/java -XX:+UseThirdPartyHeap -XX:MetaspaceSize=1G -XX:+UnlockDiagnosticVMOptions -Xint -Xmx60M -Xms60M -jar /mnt/dacapo/dacapo-evaluation-git-f480064.jar fop

caizixian committed:
Pass all three arguments in barrier slow path

fop working with interpreter and c1, but not c2

MMTK_PLAN=GenCopy /mnt/jdk/build/linux-riscv64-server-slowdebug/jdk/bin/java -XX:+UseThirdPartyHeap -XX:MetaspaceSize=1G -XX:+UnlockDiagnosticVMOptions -Xint -Xmx60M -Xms60M -jar /mnt/dacapo/dacapo-evaluation-git-f480064.jar fop

MMTK_PLAN=GenCopy /mnt/jdk/build/linux-riscv64-server-slowdebug/jdk/bin/java -XX:+UseThirdPartyHeap -XX:MetaspaceSize=1G -XX:+UnlockDiagnosticVMOptions -XX:TieredStopAtLevel=1 -Xmx60M -Xms60M -jar /mnt/dacapo/dacapo-evaluation-git-f480064.jar fop

MMTK_PLAN=GenCopy /mnt/jdk/build/linux-riscv64-server-slowdebug/jdk/bin/java -XX:+UseThirdPartyHeap -XX:MetaspaceSize=1G -XX:+UnlockDiagnosticVMOptions -Xcomp -XX:-TieredCompilation -Xmx60M -Xms60M -jar /mnt/dacapo/dacapo-evaluation-git-f480064.jar fop

#  Internal Error (/home/zixianc/mmtk-riscv/jdk-mmtk/src/hotspot/share/opto/matcher.cpp:1791), pid=2110, tid=2134
#  assert(C->node_arena()->contains(s->_leaf) || !has_new_node(s->_leaf)) failed: duplicating node that's already been matched

caizixian committed:
Fix write barrier slow path condition

fop working with c1 and interpreter

caizixian committed:
Compile with jdk-20+20

caizixian committed:
Workaround the C2 bug in codegen

caizixian committed:
Fix for openjdk/jdk#10038

caizixian committed:
Merge remote-tracking branch 'upstream/master' into jdk-20+20-mmtk

caizixian committed:
Make object barrier compile with zero

caizixian committed:
Fix for jdk-21+24

caizixian committed:
Fix object scanning

caizixian committed:
Merge tag 'v0.16.0' into jdk-21+24-mmtk-merge

caizixian committed:
Merge tag 'v0.17.0' into jdk-21+24-mmtk-merge

caizixian committed:
Merge tag 'v0.18.0' into jdk-21+24-mmtk-merge

caizixian committed:
Use stable Rust toolchain

caizixian committed:
Merge with mmtk/mmtk-openjdk@master (#234)

ArberSephirotheca committed:
Add armv8 support in the binding for mmtk/jdk:jdk-21%2B24-mmtk (#249)

Signed-off-by: ArberSephirotheca <sephirotheca17@gmail.com>

caizixian committed:
Update commit hash for mmtk/jdk#1

caizixian committed:
Fix eden_allocate for aarch64

caizixian committed:
Correctly cross compile zero variants

caizixian committed:
Fix build for riscv64

caizixian committed:
Fix for when not cross compiling

wks committed:
Update reserved argument area size for MMTk

OpenJDK no longer reserves 4 words for every frame, but instead let each
GC barrier set decide the reserved size.  We reserve 3 words for MMTk's
C1 barrier set.

See: https://github.com/openjdk/jdk/commit/36a0822633909e03656159245bbeb954efafa2f2

wks committed:
Merge branch 'master' into jdk-21+35-mmtk

wks committed:
Fix compilation errors

wks committed:
Fix mutex locking

wks committed:
Revert "Workaround the C2 bug in codegen"

This reverts commit 6086363745984732df55d02308a642ffa740a789.

wks committed:
Support compressed oops

wks committed:
Fix CI scripts for OpenJDK 21

This changes paths build options, etc., to make the CI scripts work with
OpenJDK 21.

We now use a single source of truth for the OpenJDK path instead of
defining OPENJDK_PATH in multiple scripts.

wks committed:
Rename KlassID to KlassKind

This follows the upstream change

wks committed:
Update running script for DaCapo Chopin

Use jdk21 workarounds for Cassandra and H2O

wks committed:
Merge branch 'master' into jdk-21+35-mmtk

wks committed:
Update DaCapo Chopin to 23.11-MR2

wks committed:
Merge branch 'master' into jdk-21+35-mmtk

wks committed:
Fix extended binding test

wks committed:
Fix running script

Point to the newer version of dacapo

wks committed:
Fix benchmark suite name

We just use the name defined in running-ng, i.e. "dacapochopin".  This
allows us to use modifiers such as "dacapochopin_jdk21" which expects
the benchmark suite name "dacapochopin".

wks committed:
Force all files to end with newline.

Also fixed ci-style.sh and ci-check-lineends.sh to fail if any file has
line-end problems.

wks committed:
Update README

Update contents for OpenJDK 21 and recent changes in mmtk-openjdk.

wks committed:
Merge branch 'master' into jdk-21+35-mmtk

wks committed:
Merge branch 'master' into update/master-to-jdk-21/20250815

wks committed:
Remove "other" reference

OpenJDK 21 no longer has `REF_OTHER` in the enumeration type
`ReferenceType`.

wks committed:
Merge branch 'master' into update/master-to-jdk-21/20250815

wks committed:
Fix object layout change for jdk-21.0.9+3

wks committed:
Control barrier fast paths via env var

This makes it easier to debug

wks committed:
Silently ignore `object_iterate` calls.

The MMTkHeap instance will receive calls to `object_iterate` when
connected from VisualVM and panic.  Leaving this function blank will not
properly implement the iteration, but will allow the VM to be profiled
with VisualVM.

wks committed:
Call BarrierSet::on_thread_attach

The super class method `BarrierSet::on_thread_attach` now does
initializations related to nmethods, particularly the "nmethod disarmed
guard value".

wks committed:
Merge commit 'f61d6abf6c060b498c5c6b91ec51c0f9539ac665' into jdk-21.0.9+3-mmtk

wks committed:
Merge branch 'master' into jdk-21.0.9+3-mmtk

wks committed:
Merge branch 'master' into jdk-21.0.9+3-mmtk

wks committed:
Remove debug message

wks committed:
Bump mmtk-core and deps version

wks committed:
Fix oop relocations in GC stage

wks committed:
Merge branch 'master' into jdk-21.0.9+3-mmtk

wks committed:
Merge branch 'master' into update/merge-2025-11-07

wks committed:
Minor changes for barriers

This commit addresses minor issues after merging the SATB barrier and
the barrier refactoring to the OpenJDK 21 binding.

In x86 barrier assembler, we use `testptr` for testing if a register is
zero.

OpenJDK 21 now gives 3 tmp registers to assemblers.  We use them
instead.  We also removed the workaround for OpenJDK 11 where the given
`tmp1` and `tmp2` registers overlap with operand registers.

We updated `MMTkSATBBarrierSetC2::load_at_resolved` by copying some code
from `G1BarrierSetC2`.

wks committed:
WIP: Fix compilation for AArch64.

wks committed:
Change OpenJDK repo and dir names back to `openjdk`

We revert to our old repository and directory name `openjdk` instead of
`jdk`.

-   The upstream OpenJDK project creates one repository for every major
    release, such as `jdk21u`, and updates to that major versions are
    pushed into that repository.  The `jdk` repository is for the main
    line development.  Since our `openjdk` repo actually follows update
    repos (such as `jdk21u`), it doesn't make sense to use `jdk` as the
    repository name.
-   There are many other JVM implementations out there, such as JikesRVM
    and OpenJ9.  Using the repository name `openjdk` emphasizes that it
    is OpenJDK.

qinsoon commented:
Would it be cleaner if we rename `master` to `jdk-11.x.x+x-mmtk`, and keep the branch `jdk-21.0.9+3-mmtk` and make it the default branch? We would change places where we assume `master` is the default branch in our CI.

wks commented:
> Would it be cleaner if we rename `master` to `jdk-11.x.x+x-mmtk`, and keep the branch `jdk-21.0.9+3-mmtk` and make it the default branch? We would change places where we assume `master` is the default branch in our CI.

This approach has the advantage that each branch maps uniquely to an `openjdk` repo branch, and we can still let the mmtk-core CI choose which branch to test against.

But there is an obvious disadvantage.  We may switch the default OpenJDK branch from time to time, such as from `jdk-21+35-mmtk` to `jdk-21.0.9+3-mmtk` for reasons mentioned [here](https://github.com/mmtk/mmtk-openjdk/pull/307#issuecomment-3551727574).  If we don't have a canonical `master` branch, some of us may switch the default branch while other developers are not aware of it.  They may still have `jdk-21+35-mmtk` in their local worktrees.  `git pull` won't fix that, and they need to explicitly run `git checkout`, and someone needs to tell them to do so.  This complicates things.

Another disadvantage is that we will need to change the mmtk-core CI every time we make this change, even if the CI scripts in the new branch is 100% compatible with the mmtk-core CI scripts.

So I suggest we still keep one unique `master` branch which is pointing to OpenJDK11 now, and will point to the revision of this PR in the future.  Other developers on the `master` branch can simply run `git pull` and it will be updated to the binding for OpenJDK 21.

Another advantage of keeping the `master` branch is that when we start developing a branch of a new OpenJDK version, such as `jdk-25+36-mmtk` (where `jdk-25+36` is the latest released OpenJDK LTS version now), we can develop it in parallel with the `master` branch, and we always know that `master` is the branch which `jdk-25+36-mmtk` should merge with (or rebase onto) from time to time.  When it is the time to merge that into `master`, we can still use the regular pull-request workflow for code review, as we are doing now (and previously in https://github.com/mmtk/mmtk-openjdk/pull/307).

qinsoon commented:
> We may switch the default OpenJDK branch from time to time, such as from jdk-21+35-mmtk to jdk-21.0.9+3-mmtk for reasons mentioned https://github.com/mmtk/mmtk-openjdk/pull/307#issuecomment-3551727574. If we don't have a canonical master branch, some of us may switch the default branch while other developers are not aware of it.

We could have different default branches for each JDK version, like `jdk-11`, `jdk-21` -- those are kept updated to the latest version we support.

> Another disadvantage is that we will need to change the mmtk-core CI every time we make this change, even if the CI scripts in the new branch is 100% compatible with the mmtk-core CI scripts.

I don't see this as a disadvantage. We can easily change what branch mmtk-core tests with: https://github.com/mmtk/mmtk-core/blob/e25ad8bb969f5d650de3f1946582075923bafcc9/.github/workflows/pr-binding-refs.yml#L125

Explicitly switching to a different branch is not a bad idea.

> So I suggest we still keep one unique master branch which is pointing to OpenJDK11 now, and will point to the revision of this PR in the future.

Consider this scenario: someone is using `master` with JDK 11. They did a pull, and suddenly there is a ton of conflicts and the new commit also no longer works for 11.

qinsoon commented:
My intuition is that this pull request does not really look like a normal pull request. So maybe it should not be a pull request, and we should consider other approaches to get the changes. I am happy to discuss more on this.

wks commented:
> We could have different default branches for each JDK version, like `jdk-11`, `jdk-21` -- those are kept updated to the latest version we support.

This sounds good to me, too.  We can use `jdk-21` as the default branch and update it as we update the `mmtk/openjdk` repo, and it is relatively stable.  We will switch to `jdk-25` if we feel the time is appropriate, and we will announce in our group meeting or on Zulip.

> I don't see this as a disadvantage. We can easily change what branch mmtk-core tests with: https://github.com/mmtk/mmtk-core/blob/e25ad8bb969f5d650de3f1946582075923bafcc9/.github/workflows/pr-binding-refs.yml#L125
> 
> Explicitly switching to a different branch is not a bad idea.

OK.  It's OK as long as it is easy to do so.

> Consider this scenario: someone is using `master` with JDK 11. They did a pull, and suddenly there is a ton of conflicts and the new commit also no longer works for 11.

Yes.  This is indeed a concern.  But to support multiple parallel branches like `jdk-11` and `jdk-21`, we need to phase out the `master` branch in a graceful way.  In Git, it feels abruptive to just let a branch disappear.  People usually don't expect a branch to suddenly disappear on the server.

I think one way to solve it is to update `master` to JDK 21, and introduce a branch `jdk-11` which keeps the state of `master` before the merge.  This has two implications:

-   `jdk-11` is a legacy branch, and people should generally not work on legacy JDK versions.
-   If some developers do need to work on legacy versions for some reasons, they should explicitly base on the `jdk-XX` branches.  The large number of merge conflicts is an indication that some non-trivial changes are happening, and they should switch branches locally.

Git also has a feature called "symbolic refs".  As suggested in [this page](https://stackoverflow.com/questions/14365946/create-a-branch-alias), we can use it to make aliases of branches.  We can make an actual `jdk-21` branch, and make `master` a symbolic ref to `jdk-21`.  I'll do some experiments locally on temporary repositories to see if this idea can work for mmtk-openjdk.

wks commented:
> My intuition is that this pull request does not really look like a normal pull request. So maybe it should not be a pull request, and we should consider other approaches to get the changes. I am happy to discuss more on this.

One way to make the transition is to keep an explicit `jdk-21` branch (and maybe also using `git symbolic-ref` to make `master` an alias to it), and make a PR in `mmtk-core` so that it tests against `mmtk-openjdk` branch `jdk-21`.  As for development,

-   We announce in the meeting when we formally make the transition, and/or we make `master` point to the revision (be an alias of) `jdk-21` so that we will make new changes there by default.
-   Development on `mmtk-core` will continue as usual, but new PRs on `mmtk-core` will be tested against `jdk-21`.

We should probably discuss this in our next meeting.

wks committed:
Update benchmark suite name and version

The benchmark name defined in base.yml is simply "dacapochopin", and we
reference that in normal-heap.yml that.

We also mention the MR2 version of the DaCapo Chopin JAR file wherever
appropriate.

qinsoon commented:
Kunshan and I discussed more on this.

There are a few things we agreed on:
* We will have different branches for each JDK version, such as `jdk-11`, `jdk-21`, etc.
* We usually start developing for new versions, based on the current version (e.g. developing `jdk-25` based on `jdk-21`). Because of this, we will need to merge from a old version (current trunk) to a new version. But we should avoid merging from a new version back to a old version -- we backport necessary changes by cherry-picking.

We haven't agreed on whether we should keep using `master`.

I am favoring deprecating `master`. We do not use `master` any more, and set `jdk-21` as the default branch for GitHub. The argument is that between versions, there are significant changes, and people care about the version they are using and don't want to get changes about new versions by just using `git pull`. It will be disruptive for now to deprecate `master`, as users all track our `master`. But once we deprecate `master` and ask users to track their own version branch (e.g. `jdk-21`), in the future version there is no problem.

Kunshan favors keeping `master`. Once the `jdk-21` branch is ready, we should make `master` an alias to `jdk-21`. `master` represents where our development is happening. Users will always get new changes from `master` when they do `git pull`. In future versions, we will keep changing `master` to point to the latest version we are developing. 

@wks If I missed any point from our discussion, just add to this thread.

qinsoon commented:
> Kunshan favors keeping `master`. Once the `jdk-21` branch is ready, we should make `master` an alias to `jdk-21`. `master` represents where our development is happening. Users will always get new changes from `master` when they do `git pull`. In future versions, we will keep changing `master` to point to the latest version we are developing.

I feel this is okay to me if:
* Git alias works (`git symbolic-ref`) so we don't have to push new commits to two separate branches.
* We advise users to track their corresponding version branch (`jdk-11`, `jdk-21`, etc.) if they only care about a specific version.

wks commented:
@qinsoon has summarized what we discussed today.  I'll clarify, in terms of Git commit graph, what will happen after we make the transition

This is the current status.  `master` is for OpenJDK 11.  `master` is also a proper ancestor of the `jdk-21` branch because I have merged `master` into `jdk-21` from time to time and resolved conflicts.

![mmtk-openjdk-21-branching-1](https://github.com/user-attachments/assets/ba6e4441-9701-4577-8306-750aa63d5d11)

The figure below is the state right after merging.  Before merging, we will create a branch `jdk-11` to point to the commit of the current `master`.  Because `jdk-21` is already a proper descendant of `master`, this merge will be a fast-forwarding merge, and the `jdk-21` branch stays where it was.  We will also change `master` so that it is a symbolic reference (alias) of `jdk-21` so they always point to the same commit.  The only difference between my proposal and @qinsoon's proposal is that in @qinsoon's proposal, the `master` branch will disappear rather than becoming the alias of `jdk-21`.

*(**Update**: I thought we needed a merge commit, but since it is a fast-forward merge, a merge commit is not needed.  The only things that will change are (1) `jdk-11` will disappear, and (2) `master` will move (or disappear)).*

![mmtk-openjdk-21-branching-2-fixed](https://github.com/user-attachments/assets/74175bc0-0d8b-4af0-968c-e78bccbd3c29)

The figure below is the state in the future.  Major development efforts will be on `master` (alias of `jdk-21`).

We will also have a `jdk-25` branch when OpenJDK 25 is stable.  It will start from a commit in `master` (alias of `jdk-21`), and we will make JDK-25-specific commits on the `jdk-25` branch.  From time to time, we will synchronize `jdk-25` with `master` by running `git merge master` on the `jdk-25` branch.  Eventually, `master` will become an alias of `jdk-25` (if we decide to keep `master`), and we will start the development cycle (for example, of JDK 29, or whatever will become the LTS version) on a new branch, making `jdk-21` legacy, too.

We will also have a `jdk-11` branch for the legacy OpenJDK 11 version.  We will keep it running by back-porting important features or fixes from the `master` branch to `jdk-11`.  But `jdk-11` will never merge with `master` again.

![mmtk-openjdk-21-branching-3](https://github.com/user-attachments/assets/e90c19ff-d0be-4bf0-a740-d6689e0e30df)

wks commented:
Here is a summary of some opposing ideas.

Do people stick to a particular JDK version, or do they move to the newest JDK version?

-   **stay**: People tend to stay at a particular JDK version.  This is especially true if
    +   They made significant changes to the OpenJDK runtime.  For example, they changed OpenJDK 11 to add support for publishing to support Iso.
    +   They made changes to mmtk-core and have not merged, yet.  For example, they changed the reference processing API, and also changed a large portion of the OpenJDK 11 binding.  It takes time to port the change to OpenJDK 21.
-   **move**: People tend to follow the latest JDK version we support.
    +   For GC algorithm developers, they need a stable test ground for the GC algorithm, but they don't care whether it is OpenJDK 11 or OpenJDK 21, as long as it is maintained.
    +   We mmtk-core team may develop new features that are only supported on OpenJDK 21 (and we deem it unworthy to backport to OpenJDK 11), which greatly improves the performance of any GC algorithm.

Do we use GitHub's default branch to point to the active branch, or do we have a `master` that moves from 11 to 21 to 25 to ...?

-   **GitHub default**: We set the default branch of the https://github.com/mmtk/mmtk-openjdk repo to the most actively maintained branch.
    +   Good: Whenever a developer does `git pull`, it always pulls a revision for the same OpenJDK version.
    +   Bad: Developer may stay at a branch (e.g. `jdk-11`) for a long time (e.g. 6 months) without noticing that we changed the default branch.  (Note that `git pull` doesn't notify them about default branch changes.)  They may notice the development becomes less active (because it is legacy), and they will only come to our web site or Zulip platform to find out something changed if they are curious enough.
-   **Moving `master` branch**: Have a `master` branch that moves as we change the default branch.
    +   Good: Whenever a developer does `git pull`, it automatically moves on to the most actively developed OpenJDK version.
    +   Bad: Developers that made changes to the `mmtk-openjdk` repository may suddenly see lots and lots of merge conflicts when `master` changes from one OpenJDK version to the next.
        *   Not that bad: The merge conflicts are a signal that they should either switch the remote branch to a legacy branch (e.g. `jdk-11`) or move on to the next OpenJDK version by resolving the merge conflicts.

Note that we can have both versioned branches such as `jdk-11`, `jdk-21`, `jdk-25`, etc. and a `master`, and allow the user to select whether they want to follow one particular OpenJDK version or the most actively developed version.

wks commented:
Git symbolic refs doesn't work as I expected.

If I do `git symbolic-ref refs/heads/foobarbaz refs/heads/jdk-21`, it will create a file `.git/refs/heads/foobarbaz` with content `ref: refs/heads/jdk-21`.  But I can't check out `foobarbaz`.  If I execute `git checkout foobarbaz`, the current branch will become `jdk-21`.

I can push `foobarbaz` to a remote repository by `git push origin foobarbaz:foobarbaz`.  But as a result, `refs/heads/foobarbaz` on `origin` will be an ordinary branch, not a symbolic ref.  And it will point to the same commit as `jdk-21`.  Then `foobarbaz` and `jdk-21` will behave as two independent branches.  If another user clones the `origin`, then `jdk-21` and `foobarbaz` will still be two independent branches.

I think symbolic refs is a ["plumbing" command](https://git-scm.com/book/en/v2/Git-Internals-Plumbing-and-Porcelain) supposed to be used to implement higher-level commands such as `git commit`, but not as a general means of making branch aliases.

qinsoon commented:
> Git symbolic refs doesn't work as I expected.

If we can't find a decent way to make alias of branches, we probably should just deprecate `master`.

Set aside how we deal with `master`, we could start allowing both jdk-11 and jdk-21:
1. Create a branch from the current `master` as `jdk-11`.
2. Create a branch from the current `jdk-21.0.9+3-mmtk` as `jdk-21`.
3. Set up Github branch protection for those branches.
4. (optional) Set up mergify to help with backporting (see https://github.com/mmtk/mmtk-julia/blob/master/.github/mergify.yml for an example).
5. Change `mmtk-core` CI configs to use `jdk-11` instead of `master`, including the performance tests.
6. Introduce `mmtk-core` CI configs to also test `jdk-21` for correctness.

Then we could start transitioning into `jdk-21`:
1. Use `jdk-21` for performance tests in `mmtk-core` (starting a new epoch)
1. Set `jdk-21` as the default branch on Github (if we decide to deprecate `master`).

We would also need to consider which jdk version we should include in future releases. I assume we should include `jdk-21`, starting from the next release. The question is when we exclude `jdk-11`.
